servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ConnectionContextSocketOptionTest.java

```diff
@@ -0,0 +1,244 @@
+import io.servicetalk.http.api.HttpProtocolConfig;
+import io.servicetalk.http.api.HttpResponse;
+import io.servicetalk.http.api.HttpServerBuilder;
+import io.servicetalk.http.api.SingleAddressHttpClientBuilder;
+import io.servicetalk.transport.api.HostAndPort;
+import io.servicetalk.transport.api.ServerContext;
+import io.servicetalk.transport.api.ServiceTalkSocketOptions;
+
+import org.hamcrest.Matcher;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.net.SocketOption;
+import java.net.StandardSocketOptions;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.http.api.HttpResponseStatus.OK;
+import static io.servicetalk.http.netty.HttpProtocolConfigs.h1Default;
+import static io.servicetalk.http.netty.HttpProtocolConfigs.h2Default;
+import static io.servicetalk.transport.netty.internal.AddressUtils.localAddress;
+import static io.servicetalk.transport.netty.internal.AddressUtils.serverHostAndPort;
+import static java.lang.String.valueOf;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+@RunWith(Parameterized.class)
+public class ConnectionContextSocketOptionTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConnectionContextSocketOptionTest.class);
+
+    private static final String SOCKET_OPTION_NAME = "socket-option-name";
+    private static final String SOCKET_OPTION_TYPE = "socket-option-type";
+    private static final String SOCKET_OPTION_VALUE = "socket-option-value";
+
+    private enum Protocol {
+
+        HTTP_1_1(h1Default(), false),
+        HTTP_2(h2Default(), true);
+
+        final HttpProtocolConfig config;
+        final boolean autoRead;
+
+        Protocol(HttpProtocolConfig config, boolean autoRead) {
+            this.config = config;
+            this.autoRead = autoRead;
+        }
+    }
+
+    @Rule
+    public final Timeout timeout = new ServiceTalkTestTimeout();
+
+    private final Protocol protocol;
+
+    public ConnectionContextSocketOptionTest(Protocol protocol) {
+        this.protocol = protocol;
+    }
+
+    @Parameters(name = "protocol={0}")
+    public static Object[] data() {
+        return Protocol.values();
+    }
+
+    @Test
+    public void tcpStandardSocketOptionIsNotNull() throws Exception {
+        testSocketOption("TCP_NODELAY", Boolean.class, is(notNullValue()),
+                // HTTP_2 stream channel does not have TCP_NODELAY option
+                protocol == Protocol.HTTP_2 ? equalTo("null") : not(equalTo("null")));
```

Comment:
Server connection context uses child-stream for h2 that does not allow users to query multiple TCP options for the original parent connection, like `TCP_NODELAY` / `SO_RCVBUF` / `SO_SNDBUF` / etc.
Do we want to keep it as-is? We can expose options for the original TCP connection by keeping ref for the parent's `ChannelConfig`.

Reply:
Are there any options for H2 that are applicable for a stream? If not, should we always just query the options from the parent context?

Reply:
From what I see in netty, it creates a new `DefaultChannelConfig` for each stream. And these config is a fresh new and is not related to the parent's `ChannelConfig`. Therefore, each channel has its own subset of options. In theory (but not in the current ST) some code may modify that options for each stream independently (like `AUTO_READ, `AUTO_CLOSE`, etc.) and it will not have side-effects for parent channel.

I think it will be more practical for users to be interested in parent channel options. It will require one more reference in `NettyHttpServerConnection`. Let me try and will see the overhead.

Reply:
Done: https://github.com/apple/servicetalk/pull/950/commits/4fbb1f9e2a8430c1baab762f1e16217c842875a3